File: src/memlore/application/commands/create_lore.py
from __future__ import annotations

from dataclasses import dataclass

from memlore.application.ports.clock import Clock
from memlore.application.ports.unit_of_work import UnitOfWork
from memlore.domain.exceptions import ValidationError
from memlore.domain.models.audit_record import AuditRecord
from memlore.domain.models.enums import AuditAction, KnowledgeOrigin
from memlore.domain.models.evidence import EvidenceReference
from memlore.domain.models.lore_entry import LoreEntry
from memlore.domain.models.scope import Scope


@dataclass(frozen=True, slots=True)
class CreateLoreCommand:
    statement: str
    scope: Scope
    actor_id: str
    evidence: tuple[EvidenceReference, ...] = ()


class CreateLoreHandler:
    def __init__(self, uow: UnitOfWork, clock: Clock) -> None:
        self._uow = uow
        self._clock = clock

    def handle(self, command: CreateLoreCommand) -> LoreEntry:
        actor = command.actor_id.strip()
        if not actor:
            raise ValidationError("actor must be non-empty")

        now = self._clock.now()
        entry = LoreEntry(
            statement=command.statement,
            scope=command.scope,
            created_by=actor,
            origin=KnowledgeOrigin.HUMAN_AUTHORED,
            evidence=list(command.evidence),
            created_at=now,
            updated_at=now,
        )
        audit = AuditRecord(
            target_id=entry.id,
            action=AuditAction.CREATE,
            actor_id=actor,
            created_at=now,
        )
        self._uow.lore_entries.add(entry)
        self._uow.audits.add(audit)
        self._uow.commit()
        return entry

File: internal/application/commands/create_lore.go
package commands

import (
	"context"
	"strings"

	"github.com/memlore/memlore/internal/application/ports"
	"github.com/memlore/memlore/internal/domain"
)

// CreateLoreCommand stores a human-authored lore entry.
type CreateLoreCommand struct {
	Statement string
	Scope     domain.Scope
	ActorID   string
	Evidence  []domain.EvidenceReference
}

// CreateLoreHandler handles lore creation.
type CreateLoreHandler struct {
	begin ports.UnitOfWorkFactory
	clock ports.Clock
}

func NewCreateLoreHandler(begin ports.UnitOfWorkFactory, clock ports.Clock) *CreateLoreHandler {
	return &CreateLoreHandler{begin: begin, clock: clock}
}

func (h *CreateLoreHandler) Handle(ctx context.Context, cmd CreateLoreCommand) (domain.LoreEntry, error) {
	actor := strings.TrimSpace(cmd.ActorID)
	if actor == "" {
		return domain.LoreEntry{}, &domain.ValidationError{Message: "actor must be non-empty"}
	}

	uow, err := h.begin(ctx)
	if err != nil {
		return domain.LoreEntry{}, err
	}
	defer uow.Rollback(ctx)

	now := h.clock.Now()
	entry, err := domain.NewLoreEntry(domain.NewLoreEntryInput{
		Statement: cmd.Statement,
		Scope:     cmd.Scope,
		CreatedBy: actor,
		Evidence:  cmd.Evidence,
		Now:       now,
	})
	if err != nil {
		return domain.LoreEntry{}, err
	}
	audit, err := domain.NewAuditRecord(domain.NewAuditRecordInput{
		TargetID:  entry.ID,
		Action:    domain.AuditActionCreate,
		ActorID:   actor,
		CreatedAt: now,
	})
	if err != nil {
		return domain.LoreEntry{}, err
	}

	if err := uow.LoreEntries().Add(ctx, entry); err != nil {
		return domain.LoreEntry{}, err
	}
	if err := uow.Audits().Add(ctx, audit); err != nil {
		return domain.LoreEntry{}, err
	}
	if err := uow.Commit(ctx); err != nil {
		return domain.LoreEntry{}, err
	}
	return entry, nil
}
